### ISSUE TITLE
use_build_context_synchronously is triggered when the BuildContext is used as an argument to a function 

### ISSUE THREAD

**Issue by navaronbracke** [dart-lang/linter#4784]
**Describe the issue**
The lint `use_build_context_synchronously` is triggered when passing the BuildContext to a method, after an async gap.
The context itself is not being used, though.

**To Reproduce**
```
import 'package:flutter/material.dart';

void _onButtonClick(BuildContext context) async {
  try {
    await Future.value(0);
  } catch (err) {
    SnackBarUtil.show(context, err.toString()); // lint
  }
}

class SnackBarUtil {
  static void show(BuildContext context, String msg) {
    if (context.mounted) {
      ScaffoldMessenger.of(context).showSnackBar(
        SnackBar(content: Text(msg)),
      );
    }
  }
}
```

**Expected behavior**
The lint should not trigger if the BuildContext is used as an argument. 

**Additional context**
This reproduces on Flutter 3.13.7 / Dart 3.1.3 on Dartpad

**Comment by srawlins** [comment]
@goderbauer WDYT? I think this is working as intended.

**Comment by goderbauer** [comment]
This does seem to be working as intended. @navaronbracke What's your motivation for not linting?

**Comment by matanshukry** [comment]
@goderbauer first I'll say I was the one who asked the question / reported the issue on Discord. Two points here:

1. I'm not sure if that specific behavior is a bug or not, but the "goal" of it is wrong. The point of the rule is to make sure developers won't try to use `BuildContext` when it's unavailable. Either use it without async, or check the `mounted` property. That check is done inside a method, so the call is safe - but the warning will still appear. So while it may not be a bug, the result is misleading/wrong.
2. My actual question was how to resolve the "bigger issue". That is, I have the method `bar` as a shared method, and it's called in 50 different places. I don't want to disable the rule for the entire project, nor for each of those 50 lines. What I asked for, is a way to disale it for the **callers** of that method. Something such as:

```
// ignore_for_callers: use_build_context_synchronously
void bar(BuildContext context) {
  if (context.mounted) {
    // do stuff.
  }
}
```

That way, each caller of `bar` would ignore that rule.
Not sure if something like that exists, or if a new feature request is the correct way to go though, or perhaps a differnt way.

**Comment by bwilkerson** [comment]
Because of the way that the analyzer works, it wouldn't be able to see the ignore comment at the call sites. We could potentially change that, but the cost of doing so probably outweighs the benefit.

If we wanted to provide a mechanism like this we'd need to use an annotation to do it. The annotation would basically say that it's ok to pass in an unchecked / unmounted `BuildContext`. Not only would this allow the function to be invoked without checking the state of the context, but it would also allow us to check that the context wasn't used in the annotated method without first guarding its use with an invocation of `mounted`.

**Comment by matanshukry** [comment]
@bwilkerson sure, annotation would also work from my perspective.

**Comment by navaronbracke** [comment]
The user who raised the original discussion wasn't entirely sure if the lint was working in this case.

Their original intent was to have a way to ignore the warning easily for the use of the method within the catch block.

We discussed if an ignore would fix it, but applying the ignore would be tedious.
In the end, the user was gravitating to having an annotation, so that they could indicate which method invocation can be ignored by the lint.

Personally, I do agree that the lint is working as intended.
And indeed, having the lint perform more flow analysis to determine safe usage, is too much work for what it is worth.

**Comment by srawlins** [comment]
I think we could support an annotation like `@thisFunctionGuardsAccessToItsBuildContextParameter`; the linter could check for the presence of this annotation and not report when a BuildContext is passed to the function, un-guarded.

That annotation however, should not live in the analyzer; I think it would be something provided by Flutter, if we chose to add it.
